File: Ontology/DataComplementOf.cs
namespace Ontology
{
    public class DataComplementOf: IDataComplementOf
    {
        private readonly IDataRange _dataRange;

        public DataComplementOf(
            IDataRange dataRange
            )
        {
            _dataRange = dataRange;
        }

        IDataRange IDataComplementOf.DataRange => _dataRange;

        bool IDataRange.HasMember(
            object value
            ) => !_dataRange.HasMember(value);
    }
}

File: Web/ClientApp/src/app/Ontology/DataComplementOf.ts
import { DataRange } from "./DataRange";
import { IDataComplementOf } from "./IDataComplementOf";
import { IDataRange } from "./IDataRange";
import { IDataRangeSelector } from "./IDataRangeSelector";

export class DataComplementOf
    extends DataRange
    implements IDataComplementOf
{
    constructor(
        public DataRange: IDataRange
        )
    {
        super();
    }

    HasMember(
        value: any
        ): boolean
    {
        return !this.DataRange.HasMember(value);
    }

    Select<TResult>(
        selector: IDataRangeSelector<TResult>
        ): TResult
    {
        return selector.DataComplementOf(this);
    }
}
